core/inspector.rs

```diff
@@ -315,89 +309,53 @@ impl JsRuntimeInspector {
-              );
-
-              if let Poll::Ready(Some(session_stream_item)) =
-                session.poll_next_unpin(cx)
-              {
-                let (v8_session_ptr, msg) = session_stream_item;
-                InspectorSession::dispatch_message(v8_session_ptr, msg);
+            loop {
+              // Accept new connections.
+              let poll_result = sessions.session_rx.poll_next_unpin(cx);
+              match poll_result {
+                Poll::Ready(Some(session_proxy)) => {
+                  let session = InspectorSession::new(
+                    self.v8_inspector.clone(),
+                    session_proxy,
+                    false,
+                  );
+                  sessions.established.push(session);
+                  continue; // Poll `session_rx` until it returns `Pending`.
+                }
+                Poll::Ready(None) => unreachable!(), // Is it really unreachable?.
+                Poll::Pending => {}
               }
-              sessions.established.push(session);
-            }
 
-            loop {
+              // Poll all established sessions for incoming messages, and dispatch
+              // them to the V8 inspector.
               let poll_result = sessions.established.poll_next_unpin(cx);
               match poll_result {
                 Poll::Ready(Some(session_stream_item)) => {
                   let (v8_session_ptr, msg) = session_stream_item;
                   InspectorSession::dispatch_message(v8_session_ptr, msg);
+                  continue; // Keep polling established sessions and accepting new ones, until everything is `Pending`.
                 }
-                _ => break,
-              }
-            }
-
-            // Keep polling until a session connects and tell us to start
-            // running.
-            if self.flags.borrow().waiting_for_session {
-              Poll::Pending
-            } else {
-              Poll::Ready(())
-            }
-          }));
-          self.waker.update(|w| {
-            assert_eq!(w.poll_state, PollState::BlockOnWaitingForSession);
-            w.poll_state = PollState::SyncPolling;
-          });
-        }
-        PollState::BlockOnPause => {
-          futures::executor::block_on(futures::future::poll_fn(|cx| {
-            // Accept new connections.
-            let poll_result = sessions.session_rx.poll_next_unpin(cx);
-            if let Poll::Ready(Some(session_proxy)) = poll_result {
-              let mut session = InspectorSession::new(
-                self.v8_inspector.clone(),
-                session_proxy,
-                false,
-              );
-
-              if let Poll::Ready(Some(session_stream_item)) =
-                session.poll_next_unpin(cx)
-              {
-                let (v8_session_ptr, msg) = session_stream_item;
-                InspectorSession::dispatch_message(v8_session_ptr, msg);
+                Poll::Ready(None) => unreachable!(), // Is it really unreachable?.
```

Comment:
We also hold the `sessions` container

Reply:
I think we could still hit it when all the sessions disconnect.
We may be holding the `sessions` container, but we are calling `poll()` on those streams too, which may empty them out.